File: app/src/main/java/net/mieczkowski/yelpbusinessexample/recyclerAdapters/base/BaseViewHolder.java
package net.mieczkowski.yelpbusinessexample.recyclerAdapters.base;

import android.support.v7.widget.RecyclerView;
import android.view.View;

import butterknife.ButterKnife;

/**
 * Created by Josh Mieczkowski on 10/20/2017.
 */

public abstract class BaseViewHolder<T> extends RecyclerView.ViewHolder {

    private BaseAdapter<T> baseAdapter;

    public BaseViewHolder(View itemView) {
        super(itemView);
        ButterKnife.bind(this, itemView);

        if (adapterHandleClick()) {
            this.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    baseAdapter.onRowClick(getAdapterPosition());
                }
            });
        } else {
            this.itemView.setOnClickListener(null);
        }
    }

    void setBaseAdapter(BaseAdapter<T> baseAdapter) {
        this.baseAdapter = baseAdapter;
    }

    protected boolean adapterHandleClick() {
        return true;
    }

    public abstract void onBind(T object);

}

File: app/src/main/java/net/mieczkowski/yelpbusinessexample/recyclerAdapters/base/BaseViewHolder.kt
package net.mieczkowski.yelpbusinessexample.recyclerAdapters.base

import android.support.v7.widget.RecyclerView
import android.view.View

/**
 * Created by Josh Mieczkowski on 10/20/2017.
 */

abstract class BaseViewHolder<T>(itemView: View) : RecyclerView.ViewHolder(itemView) {

    private lateinit var baseAdapter: BaseAdapter<T>

    init {

        if (adapterHandleClick()) {
            this.itemView.setOnClickListener { baseAdapter.onRowClick(adapterPosition) }
        } else {
            this.itemView.setOnClickListener(null)
        }
    }

    fun setBaseAdapter(baseAdapter: BaseAdapter<T>) {
        this.baseAdapter = baseAdapter
    }

    protected fun adapterHandleClick(): Boolean {
        return true
    }

    abstract fun onBind(item: T?)

}